### PR TITLE
[Core] Non-historical and custom variable normal

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**📝 Description**
This extends the normal calculation utilities to make possible the usage of a custom variable to store the normal as well as to store it in the non-historical nodal database. The current behavior is kept by default so nothing should be broken. Besides, I avoided any change in the API as this class is way extended in Kratos. Tests have been added accordingly.

Last but not least, I noted that there are some methods that, according to my understanding, belong to the no-submodelparts era. IMO it's time to start deprecating these.

### CODE REVIEW

File: kratos/utilities/normal_calculation_utils.cpp [hunk 1]
```diff
@@ -24,7 +24,9 @@
 namespace Kratos
 {
 template <class TContainerType>
-void NormalCalculationUtils::CalculateNormalsInContainer(ModelPart& rModelPart)
+void NormalCalculationUtils::CalculateNormalsInContainer(
+    ModelPart& rModelPart,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 2]
```diff
@@ -45,15 +47,17 @@ void NormalCalculationUtils::CalculateNormalsInContainer(ModelPart& rModelPart)
...
 
 /***********************************************************************************/
 /***********************************************************************************/
 
-template <class TContainerType>
-void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
+template <class TContainerType, bool TIsHistorical>
+void NormalCalculationUtils::InitializeNormals(
+    ModelPart& rModelPart,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 3]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
 
 template <>
-KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::CalculateNormals<Condition>(
+KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::CalculateNormals<Condition, true>(
     ModelPart& rModelPart,
     const bool EnforceGenericGeometryAlgorithm,
-    const bool ConsiderUnitNormal
-    )
+    const bool ConsiderUnitNormal,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 4]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
-    }
-    use_simplex = rModelPart.GetCommunicator().GetDataCommunicator().MaxAll(use_simplex);
-    if (use_simplex) {
-        CalculateOnSimplex(rModelPart, dimension);
+template <>
+KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::CalculateNormals<Condition, false>(
+    ModelPart& rModelPart,
+    const bool EnforceGenericGeometryAlgorithm,
+    const bool ConsiderUnitNormal,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 5]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
 
 template<>
-KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::CalculateNormals<Element>(
+KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::CalculateNormals<Element, true>(
     ModelPart& rModelPart,
     const bool EnforceGenericGeometryAlgorithm,
-    const bool ConsiderUnitNormal
-    )
+    const bool ConsiderUnitNormal,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 6]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
 /***********************************************************************************/
 
-template<class TEntityType>
+template<class TEntityType, bool TIsHistorical>
 void NormalCalculationUtils::CalculateUnitNormals(
     ModelPart& rModelPart,
-    const bool EnforceGenericGeometryAlgorithm
-    )
+    const bool EnforceGenericGeometryAlgorithm,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 7]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
 
 /***********************************************************************************/
 /***********************************************************************************/
 
 void NormalCalculationUtils::CalculateOnSimplex(
     ConditionsArrayType& rConditions,
-    const std::size_t Dimension
-    )
+    const std::size_t Dimension,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 8]
```diff
@@ -91,117 +99,110 @@ void NormalCalculationUtils::InitializeNormals(ModelPart& rModelPart)
...
     KRATOS_CATCH("")
+}
 
+/***********************************************************************************/
+/***********************************************************************************/
+
+void NormalCalculationUtils::CalculateOnSimplexNonHistorical(
+    ConditionsArrayType& rConditions,
+    const std::size_t Dimension,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 9]
```diff
@@ -247,15 +248,34 @@ void NormalCalculationUtils::CalculateNormalShapeDerivativesOnSimplex(
 /***********************************************************************************/
 
 void NormalCalculationUtils::CalculateOnSimplex(
-    ModelPart& rModelPart)
+    ModelPart& rModelPart,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 10]
```diff
@@ -265,17 +285,35 @@ void NormalCalculationUtils::CalculateOnSimplex(
 
 void NormalCalculationUtils::CalculateOnSimplex(
     ModelPart& rModelPart,
-    const std::size_t Dimension
-    )
+    const std::size_t Dimension,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 11]
```diff
@@ -325,16 +370,16 @@ void NormalCalculationUtils::ComputeUnitNormalsFromAreaNormals(ModelPart& rModel
 
 void NormalCalculationUtils::CalculateNormal2D(
     Condition& rCondition,
-    array_1d<double,3>& rAn
-    )
+    array_1d<double,3>& rAn,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 12]
```diff
@@ -375,8 +420,8 @@ void NormalCalculationUtils::CalculateNormal3D(
     Condition& rCondition,
     array_1d<double,3>& rAn,
     array_1d<double,3>& rv1,
-    array_1d<double,3>& rv2
-    )
+    array_1d<double,3>& rv2,
+    const Variable<array_1d<double,3>>& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const Variable<array_1d<double,3>>& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 13]
```diff
@@ -472,19 +517,19 @@ KRATOS_API(KRATOS_CORE) ModelPart::ElementsContainerType& NormalCalculationUtils
...
 /***********************************************************************************/
 
-template<class TContainerType>
+template<class TContainerType, bool TIsHistorical>
 void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
     ModelPart& rModelPart,
-    const bool ConsiderUnitNormal
-    )
+    const bool ConsiderUnitNormal,
+    const NormalVariableType& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const NormalVariableType& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 14]
```diff
@@ -515,27 +560,137 @@ void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
...
     KRATOS_CATCH("");
 }
 
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+KRATOS_API(KRATOS_CORE) array_1d<double,3>& NormalCalculationUtils::GetNormalValue<true>(
+    NodeType& rNode,
+    const NormalVariableType& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const NormalVariableType& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 15]
```diff
@@ -515,27 +560,137 @@ void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
...
+    return rNode.FastGetSolutionStepValue(rNormalVariable);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<>
+KRATOS_API(KRATOS_CORE) array_1d<double,3>& NormalCalculationUtils::GetNormalValue<false>(
+    NodeType& rNode,
+    const NormalVariableType& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const NormalVariableType& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 16]
```diff
@@ -515,27 +560,137 @@ void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
...
+{
+    return rNode.GetValue(rNormalVariable);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+KRATOS_API(KRATOS_CORE) bool NormalCalculationUtils::CheckUseSimplex(
+    const ModelPart& rModelPart,
+    const bool EnforceGenericGeometryAlgorithm)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const bool EnforceGenericGeometryAlgorithm
    )
```

File: kratos/utilities/normal_calculation_utils.cpp [hunk 17]
```diff
@@ -515,27 +560,137 @@ void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
...
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<bool TIsHistorical>
+KRATOS_API(KRATOS_CORE) void NormalCalculationUtils::AuxiliaryCalculateOnSimplex(
+    ConditionsArrayType& rConditions,
+    const std::size_t Dimension,
+    const NormalVariableType& rNormalVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.cpp]:
```suggestion
    const NormalVariableType& rNormalVariable
    )
```

File: kratos/utilities/normal_calculation_utils.h [hunk 18]
```diff
@@ -265,16 +285,16 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
     void CalculateOnSimplex(
         ModelPart& rModelPart,
         const std::size_t Dimension,
-        Variable<TValueType>& rVariable,
+        const Variable<TValueType>& rVariable,
         const TValueType Zero,
-        const double rAlpha
-        )
+        const double rAlpha,
+        const NormalVariableType& rNormalVariable = NORMAL)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        )
```

File: kratos/utilities/normal_calculation_utils.h [hunk 19]
```diff
@@ -189,17 +210,17 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
     void CalculateOnSimplex(
         ModelPart& rModelPart,
         const std::size_t Dimension,
-        Variable<TValueType>& rVariable,
-        const TValueType Zero
-        )
+        const Variable<TValueType>& rVariable,
+        const TValueType Zero,
+        const NormalVariableType& rNormalVariable = NORMAL)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        )
```

File: kratos/utilities/normal_calculation_utils.h [hunk 20]
```diff
@@ -97,45 +100,54 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      * @param rModelPart The model part to compute
      */
     template<class TContainerType>
-    void CalculateNormalsInContainer(ModelPart& rModelPart);
+    void CalculateNormalsInContainer(
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 21]
```diff
@@ -97,45 +100,54 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
      */
-    template<class TEntity>
+    template<class TEntity, bool TIsHistorical = true>
     void CalculateNormals(
         ModelPart& rModelPart,
         const bool EnforceGenericGeometryAlgorithm = false,
-        const bool ConsiderUnitNormal = false
-        );
+        const bool ConsiderUnitNormal = false,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 22]
```diff
@@ -97,45 +100,54 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
      * @tparam TEntity The entity type considered
      */
-    template<class TEntity>
+    template<class TEntity, bool TIsHistorical = true>
     void CalculateUnitNormals(
         ModelPart& rModelPart,
-        const bool EnforceGenericGeometryAlgorithm = false
-        );
+        const bool EnforceGenericGeometryAlgorithm = false,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 23]
```diff
@@ -97,45 +100,54 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
      * @param Dimension Spatial dimension (2 or 3)
+     * @param rNormalVariable Variable on which the normal values are to be stored
      * @note This function is not recommended for distributed (MPI) runs, as the user has to ensure that the calculated normals are assembled between processes. The overload of this function that takes a ModelPart is preferable in ths case, as it performs the required communication.
      */
     void CalculateOnSimplex(
         ConditionsArrayType& rConditions,
-        const std::size_t Dimension
-        );
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 24]
```diff
@@ -97,45 +100,54 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    void CalculateOnSimplexNonHistorical(
+        ConditionsArrayType& rConditions,
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 25]
```diff
@@ -157,8 +169,13 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      */
     void CalculateOnSimplex(
         ModelPart& rModelPart,
-        const std::size_t Dimension
-        );
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 26]
```diff
@@ -157,8 +169,13 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    void CalculateOnSimplexNonHistorical(
+        ModelPart& rModelPart,
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 27]
```diff
@@ -167,8 +184,12 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      * @note Use this fuction instead of its overload taking a Conditions array for MPI applications, as it will take care of communication between partitions.
      */
     void CalculateOnSimplex(
-        ModelPart& rModelPart
-        );
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 28]
```diff
@@ -167,8 +184,12 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    void CalculateOnSimplexNonHistorical(
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable = NORMAL);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 29]
```diff
@@ -247,10 +268,9 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
     void CalculateOnSimplex(
         ModelPart& rModelPart,
         const std::size_t Dimension,
-        Variable<TValueType>& rVariable
-        )
+        const Variable<TValueType>& rVariable)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const Variable<TValueType>& rVariable
        )
```

File: kratos/utilities/normal_calculation_utils.h [hunk 30]
```diff
@@ -379,16 +399,17 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
     void CalculateOnSimplexLowMemory(
         ModelPart& rModelPart,
         int Dimension,
-        Variable<TValueType>& rVariable,
-        const TValueType Zero,const double rAlpha)
+        const Variable<TValueType>& rVariable,
+        const TValueType Zero,const double rAlpha,
+        const NormalVariableType& rNormalVariable = NORMAL)
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable = NORMAL
        )
```

File: kratos/utilities/normal_calculation_utils.h [hunk 31]
```diff
@@ -499,15 +520,22 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
      * @param rModelPart The model part to compute
      * @tparam TEntity The entity type considered
+     * @tparam TIsHistorical Specifies if the historical or non-historical nodal database is used
      */
-    template<class TEntity>
-    void InitializeNormals(ModelPart& rModelPart);
+    template<class TEntity, bool TIsHistorical>
+    void InitializeNormals(
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 32]
```diff
@@ -499,15 +520,22 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
     /**
      * @brief It computes the unit normals from the area normals
      * @param rModelPart The model part to compute
+     * @tparam TIsHistorical Specifies if the historical or non-historical nodal database is used
      */
-    void ComputeUnitNormalsFromAreaNormals(ModelPart& rModelPart);
+    template<bool TIsHistorical>
+    void ComputeUnitNormalsFromAreaNormals(
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 33]
```diff
@@ -516,8 +544,8 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      */
     static void CalculateNormal2D(
         Condition& rCondition,
-        array_1d<double,3>& rAn
-        );
+        array_1d<double,3>& rAn,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 34]
```diff
@@ -538,8 +566,8 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
         Condition& rCondition,
         array_1d<double,3>& rAn,
         array_1d<double,3>& rv1,
-        array_1d<double,3>& rv2
-        );
+        array_1d<double,3>& rv2,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 35]
```diff
@@ -563,12 +591,28 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+     * @tparam TIsHistorical Specifies if the historical or non-historical nodal database is used
      */
-    template<class TContainerType>
+    template<class TContainerType, bool TIsHistorical>
     void CalculateNormalsUsingGenericAlgorithm(
         ModelPart& rModelPart,
-        const bool ConsiderUnitNormal = false
-        );
+        const bool ConsiderUnitNormal,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 36]
```diff
@@ -563,12 +591,28 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    template<bool TIsHistorical>
+    array_1d<double,3>& GetNormalValue(
+        NodeType& rNode,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 37]
```diff
@@ -563,12 +591,28 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    bool CheckUseSimplex(
+        const ModelPart& rModelPart,
+        const bool EnforceGenericGeometryAlgorithm);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const bool EnforceGenericGeometryAlgorithm
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 38]
```diff
@@ -563,12 +591,28 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+
+    template<bool TIsHistorical>
+    void AuxiliaryCalculateOnSimplex(
+        ConditionsArrayType& rConditions,
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable);
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h]:
```suggestion
        const NormalVariableType& rNormalVariable
        );
```

File: kratos/utilities/normal_calculation_utils.h [hunk 39]
```diff
@@ -563,11 +608,31 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+        NodeType& rNode,
+        const NormalVariableType& rNormalVariable
+        );
+
+    bool CheckUseSimplex(
+        const ModelPart& rModelPart,
+        const bool EnforceGenericGeometryAlgorithm
+        );
+
+    template<bool TIsHistorical>
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:707]: Doxygen

File: kratos/utilities/normal_calculation_utils.h [hunk 40]
```diff
@@ -563,11 +608,31 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
+        const NormalVariableType& rNormalVariable
+        );
+
+    template<bool TIsHistorical>
+    array_1d<double,3>& GetNormalValue(
+        NodeType& rNode,
+        const NormalVariableType& rNormalVariable
+        );
+
+    bool CheckUseSimplex(
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:694]: Doxygen

File: kratos/utilities/normal_calculation_utils.h [hunk 41]
```diff
@@ -563,11 +608,31 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
-    template<class TContainerType>
+    template<class TContainerType, bool TIsHistorical>
     void CalculateNormalsUsingGenericAlgorithm(
         ModelPart& rModelPart,
-        const bool ConsiderUnitNormal = false
+        const bool ConsiderUnitNormal,
+        const NormalVariableType& rNormalVariable
+        );
+
+    template<bool TIsHistorical>
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:665]: Doxygen

File: kratos/utilities/normal_calculation_utils.h [hunk 42]
```diff
@@ -167,7 +191,13 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      * @note Use this fuction instead of its overload taking a Conditions array for MPI applications, as it will take care of communication between partitions.
      */
     void CalculateOnSimplex(
-        ModelPart& rModelPart
+        ModelPart& rModelPart,
+        const NormalVariableType& rNormalVariable = NORMAL
+        );
+
+    void CalculateOnSimplexNonHistorical(
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:227]: Doxygen

File: kratos/utilities/normal_calculation_utils.h [hunk 43]
```diff
@@ -157,7 +174,14 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
      */
     void CalculateOnSimplex(
         ModelPart& rModelPart,
-        const std::size_t Dimension
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL
+        );
+
+    void CalculateOnSimplexNonHistorical(
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:202]: Doxygen

File: kratos/utilities/normal_calculation_utils.h [hunk 44]
```diff
@@ -119,22 +127,31 @@ class KRATOS_API(KRATOS_CORE) NormalCalculationUtils
...
      * @note This function is not recommended for distributed (MPI) runs, as the user has to ensure that the calculated normals are assembled between processes. The overload of this function that takes a ModelPart is preferable in ths case, as it performs the required communication.
      */
     void CalculateOnSimplex(
         ConditionsArrayType& rConditions,
-        const std::size_t Dimension
+        const std::size_t Dimension,
+        const NormalVariableType& rNormalVariable = NORMAL
+        );
+
+    void CalculateOnSimplexNonHistorical(
```
Comment (loumalouomega) [inline kratos/utilities/normal_calculation_utils.h:163]: Doxygen

**Review by loumalouomega — DISMISSED** [review]
Minor comments, basically Doxygen. There are some TODO for the parallel, what is the problem?

File: kratos/utilities/normal_calculation_utils.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.cpp]: Why all these changes? I've just checked our [style guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide) and there is no comment about the brackets.

**Comment by rubenzorrilla** [comment]
> Minor comments, basically Doxygen. There are some TODO for the parallel, what is the problem?

The TODOs were already there. I've just added those ones directly involved in required refactoring.

File: kratos/utilities/normal_calculation_utils.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/normal_calculation_utils.cpp]: Previous code was like that (see diff)

**Comment by loumalouomega** [comment]
> > Minor comments, basically Doxygen. There are some TODO for the parallel, what is the problem?
> 
> The TODOs were already there. I've just added those ones directly involved in required refactoring.

Okay

File: kratos/utilities/normal_calculation_utils.h (hunk 40 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.h:694]: Done.

File: kratos/utilities/normal_calculation_utils.h (hunk 42 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.h:227]: Done.

File: kratos/utilities/normal_calculation_utils.h (hunk 43 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.h:202]: Done.

File: kratos/utilities/normal_calculation_utils.h (hunk 44 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.h:163]: Done.

**Review by loumalouomega — DISMISSED** [review]

**Review by loumalouomega — APPROVED** [review]

File: kratos/utilities/normal_calculation_utils.h (hunk 41 above)
Comment (rubenzorrilla) [reply kratos/utilities/normal_calculation_utils.h:665]: Done.

**Review by loumalouomega — DISMISSED** [review]
CL is failing (maybe not related)

**Comment by rubenzorrilla** [comment]
> CL is failing (maybe not related)

I've the feeling it's not. The nightly is failing because of the `ConstitutiveLawsApplication`. The FullDebug builds because the core tests take too long...

**Review by loumalouomega — DISMISSED** [review]

**Comment by rubenzorrilla** [comment]
Finally I had some time to fix this. We had an implementation inconsistency in the calculation of the unit normals. The point is that we were normalizing all the nodal values in a model part, but we were only computing the nodal normals in those nodes belonging to a condition (or element). This is not a problem with the historical database, since memory is allocated when creating the nodes with the current data value container. However, in the non-historical case we need to do a first `SetValue` to make the allocation threadsafe. My first implementation was doing such `SetValue` only for the nodes belonging to a condition, but we were normalizing all the model part nodes. The latest commit fixes this.

@loumalouomega note that I required to change a template argument from entity value (condition/element) to container type (ConditionsContainerType/ElementsContainerType of the ModelPart). This is now used in the `ComputeUnitNormalsFromAreaNormals` in order to properly loop the container of interest. This is why there are so many changes in other files.

File: kratos/utilities/normal_calculation_utils.cpp [hunk 45]
```diff
@@ -298,24 +346,35 @@ void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
 /***********************************************************************************/
 /***********************************************************************************/
 
-void NormalCalculationUtils::ComputeUnitNormalsFromAreaNormals(ModelPart& rModelPart)
+template<class TContainerType, bool TIsHistorical>
```
Comment (rubenzorrilla) [inline kratos/utilities/normal_calculation_utils.cpp:349]: This is the template argument change I was referring to.

File: kratos/utilities/normal_calculation_utils.cpp [hunk 46]
```diff
@@ -298,24 +346,35 @@ void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
...
-        auto& r_normal = rNode.FastGetSolutionStepValue(NORMAL);
-        const double norm_normal = norm_2(r_normal);
-
-        KRATOS_ERROR_IF(rNode.Is(INTERFACE) && norm_normal <= std::numeric_limits<double>::epsilon())
-            << "ERROR:: ZERO NORM NORMAL IN NODE: " << rNode.Id() << std::endl;
-        r_normal /= norm_normal;
+    // We iterate over the nodes of the corresponding container
+    // Note that we need to iterate the corresponding entities in case there are
+    // nodes in the model part that do not belong to any entity.
+    auto& r_entity_container = GetContainer<TContainerType>(rModelPart);
```
Comment (rubenzorrilla) [inline kratos/utilities/normal_calculation_utils.cpp:368]: By adding it I can get the container of interest in here.

**Comment by rubenzorrilla** [comment]
Finally I had some time to fix this. We had an implementation inconsistency in the calculation of the unit normals. The point is that we were normalizing all the nodal values in a model part, but we were only computing the nodal normals in those nodes belonging to a condition (or element). This is not a problem with the historical database, since memory is allocated when creating the nodes with the current data value container. However, in the non-historical case we need to do a first `SetValue` to make the allocation threadsafe. My first implementation was doing such `SetValue` only for the nodes belonging to a condition, but we were normalizing all the model part nodes. The latest commit fixes this.

@loumalouomega note that I required to change a template argument from entity value (condition/element) to container type (ConditionsContainerType/ElementsContainerType of the ModelPart). This is now used in the `ComputeUnitNormalsFromAreaNormals` in order to properly loop the container of interest. This is why there are so many changes in other files.

**Review by loumalouomega — APPROVED** [review]
If my tests pass, I assume it is OK

**Pull request merged**